theano/gpuarray/c_code/ctc_wrapper.c

```diff
@@ -0,0 +1,293 @@
+    switch ( (*out_costs)->ga.typecode )
+    {
+    case GA_FLOAT:
+        costs = (float *) PyGpuArray_DEV_DATA( *out_costs );
+        break;
+    default:
+        ctc_context_destroy( context );
+        PyErr_SetString(PyExc_TypeError, "Unsupported type for costs!");
+        return 1;
+    }
+
+    float * gradients = NULL;
+
+    if ( NULL != out_gradients )  // if gradient computation is not disabled
+    {
+        if ( NULL == *out_gradients ||
+             PyGpuArray_NDIM( *out_gradients ) != 3 ||
+             PyGpuArray_DIMS( *out_gradients )[0] != PyGpuArray_DIMS( in_activations )[0] ||
+             PyGpuArray_DIMS( *out_gradients )[1] != PyGpuArray_DIMS( in_activations )[1] ||
+             PyGpuArray_DIMS( *out_gradients )[2] != PyGpuArray_DIMS( in_activations )[2] )
+        {
+            Py_XDECREF( *out_gradients );
+
+            const size_t * activation_dims = PyGpuArray_DIMS( in_activations );
+            *out_gradients = pygpu_zeros( 3, activation_dims, GA_FLOAT, GA_C_ORDER,
+                gpu_context, Py_None );
+
+            if ( NULL == *out_gradients )
+            {
+                ctc_context_destroy( context );
+
+                PyErr_Format( PyExc_MemoryError,
+                    "Could not allocate storage for CTC gradients!" );
+                return 1;
+            }
+        }
+        else
+        {
+            GpuArray_memset( &((*out_gradients)->ga), 0 );
+        }
+
+        switch ( (*out_gradients)->ga.typecode )
+        {
+        case GA_FLOAT:
+            gradients = (float *) PyGpuArray_DEV_DATA( *out_gradients );
+            break;
+        default:
+            ctc_context_destroy( context );
+            PyErr_SetString(PyExc_TypeError, "Unsupported type for gradients!");
+            return 1;
+        }
+    }
+
+    size_t gpu_workspace_size;
+    int ctc_error = 0;
+
+    ctc_error = ctc_check_result( get_workspace_size( context->label_lengths,
+        context->input_lengths, alphabet_size, minibatch_size, context->options,
+        &gpu_workspace_size ),
+        "Failed to obtain CTC workspace size!" );
+
+    if ( ctc_error )  // Exception is set by ctc_check_result, return error here
+    {
+        // Destroy previous CTC context before returning exception
+        ctc_context_destroy( context );
+
+        return 1;
+    }
+
+    context->workspace = pygpu_empty(1, &gpu_workspace_size, GA_BYTE,
+        GA_C_ORDER, gpu_context, Py_None );
+
+    if ( NULL == context->workspace )
+    {
+        ctc_context_destroy( context );
+
+        PyErr_Format( PyExc_MemoryError,
+            "Failed to allocate memory for CTC workspace!" );
+        return 1;
+    }
+
+    ctc_error = ctc_check_result( compute_ctc_loss( activations, gradients,
+        context->flat_labels, context->label_lengths, context->input_lengths,
+        alphabet_size, minibatch_size, costs, PyGpuArray_DEV_DATA(context->workspace),
+        context->options ), "Failed to compute CTC loss function!" );
```

Comment:
This should probably be wrapped in cuda_enter()/cuda_exit() to set the proper ambiant context.  I don't know if the get_workspace_size() function above also needs this.